### PR TITLE
Experimental support for outline type checking

### PR DESCRIPTION
**Pull request by retronym**
A new mode for `PipelineMain` that can shorten the critical path of a build and extract more parallelism from the post-typer compiler phases (at the cost of instantiating multiple instances of `Global`).

First, we compile all sources with `-Youtline -Ystop-after:pickler`. The first option disables typechecking of the RHS of definitions that have ascribed types.

The RHS is _not_ skipped when it contains a `super` call that might need a super accessor. (hat tip to @smarter for finding this corner case in dotty). Are there more like it that we haven't identified?

The pickles are then exported and made available to downstream compilation, as is also the case with the `strategy=Pipeline`. The sources for the current projects are then compiled in parallel in chunks, also using the exported pickles on the classpath.

Testing this out `guardian-frontend`. In the past I've experimented with the full multi-module build, but in this case I want to focus on the `common` project that is the bottleneck the build.

## Pipeline + Outline
```
$ mkdir -p /tmp/pickle-cache && ./build/pack/bin/scala -J-Dsun.reflect.inflationThreshold=0  -J-Xmx4G -Dscala.pipeline.cache.name.table=true -Dscalac.filebasedcache.defer.close.ms=30000 -Dscala.pipeline.picklecache=/tmp/pickle-cache -Dscala.pipeline.group.size=128  -Dscala.pipeline.strategy=outlinetypepipeline -Dscala.pipeline.strip.external.classpath=true -Dscala.pipeline.cache.macro.classloader=true -Dscala.pipeline.use.jar=true scala.tools.nsc.PipelineMainTest /Users/jz/code/guardian-frontend/common/target/compile.args
...
====== ITERATION 20=======
parallelism = 8, strategy = OutlineTypePipeline
Exported external classpath in 4 ms
Wrote project dependency graph to: /Users/jz/code/scala/projects.dot
/Users/jz/code/guardian-frontend/common:compile: scalac outline: start
/Users/jz/code/guardian-frontend/common:compile: scalac outline: done 1378 ms
/Users/jz/code/guardian-frontend/common:compile: scalac (1/4): start
/Users/jz/code/guardian-frontend/common:compile: scalac (2/4): start
/Users/jz/code/guardian-frontend/common:compile: scalac (4/4): start
/Users/jz/code/guardian-frontend/common:compile: scalac (3/4): start
/Users/jz/code/guardian-frontend/common:compile: scalac (4/4): done 1729 ms
/Users/jz/code/guardian-frontend/common:compile: scalac (3/4): done 1898 ms
/Users/jz/code/guardian-frontend/common:compile: scalac (1/4): done 3074 ms
/Users/jz/code/guardian-frontend/common:compile: scalac (2/4): done 4698 ms
PROGRESS: 6 / 6
 Wall Clock: 6077 ms
Chrome trace written to /Users/jz/code/scala/build-OutlineTypePipeline-20.trace
```

## Pipeline 

```
$ ... -Dscala.pipeline.strategy=pipeline ...
====== ITERATION 20=======
parallelism = 8, strategy = Pipeline
Exported external classpath in 4 ms
Wrote project dependency graph to: /Users/jz/code/scala/projects.dot
/Users/jz/code/guardian-frontend/common:compile: scalac: start
/Users/jz/code/guardian-frontend/common:compile: scalac outline: done 4232 ms
/Users/jz/code/guardian-frontend/common:compile: scalac: exported pickles 52 ms
/Users/jz/code/guardian-frontend/common:compile: scalac: done 4523 ms
PROGRESS: 3 / 3
 Wall Clock: 8807 ms
Chrome trace written to /Users/jz/code/scala/build-Pipeline-20.trace
```

## Traditional

```
$ ... -Dscala.pipeline.strategy=traditional ...
====== ITERATION 20=======
parallelism = 8, strategy = Traditional
Wrote project dependency graph to: /Users/jz/code/scala/projects.dot
/Users/jz/code/guardian-frontend/common:compile: scalac (1/1): start
/Users/jz/code/guardian-frontend/common:compile: scalac (1/1): done 9343 ms
PROGRESS: 3 / 3
Wall Clock: 9409 ms
Chrome trace written to /Users/jz/code/scala/build-Traditional-20.trace
```

## Summary

 - The main benefit of the new strategy here is reducing the "time-to-pickles" down to 1.4s (vs  9.3 for traditional vs 4.2 for non-outline pipeline)
 - The wall clock time is also reduced somewhat (6.1s vs 8.8s), which may be interesting in itself when enough cores are available. There will be rapidly diminishing returns as the chunk size gets too small, however, due to the current approach of using a Global instance per-chunk.

## Caveats

 - Incompatible with the non-local optimizations in the backend as .class files for symbols backed by exported pickles aren't available. This shouldn't be a big problem in "dev mode".
 - To get the most of out of the caching of macro classloaders, a hand crafted `-Ymacro-classpath` setting is needed with the macro implementation.

## Future work:

  - Name table sharing across concurrently running instances of the compiler: https://github.com/retronym/scala/pull/46

### CODE REVIEW

**Comment by retronym** [comment]
I'm testing this out now on some real world projects. Will mark as WIP while I do that.

**Comment by retronym** [comment]
Here's how I've exported suitable `.args` files for the Akka build and build with `PipelineMain`-s three strategies: https://github.com/retronym/akka/pull/1/files

I'm still analysing the results.

One functional problem so far:

  - [x] chome trace output is broken under this new mode.

File: test/junit/scala/tools/nsc/PipelineMainTest.scala [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+      // Bytecode should be identical regardless of compilation strategy.
+      deleteBaseAfterTest = false
+      assertDirectorySame(reference, recompiled, strategy.toString)
+      deleteBaseAfterTest = true
+    }
+  }
+
+  private lazy val allBuilds = List(m1, b2, b3, b4, b5SuperAccessor)
+
+  private lazy val m1: Build = {
```
Comment (adriaanm) [inline test/junit/scala/tools/nsc/PipelineMainTest.scala:84]: Each of these test scenarios could really use a short description of what's tested. I can mostly guess, but let's make it explicit.

**Comment by adriaanm** [comment]
Very nice! I would like to see a short motivation for each test case, but LGTM otherwise.

**Pull request merged**